packages/worker/src/worker.ts

```diff
@@ -1297,30 +1305,57 @@ export class Worker {
    * This function does not throw, it will log in case of missing sinks
    * or failed sink function invocations.
    */
-  protected async processSinkCalls(externalCalls: SinkCall[], info: WorkflowInfo, isReplaying: boolean): Promise<void> {
+  protected async processSinkCalls(externalCalls: SinkCall[], isReplaying: boolean): Promise<void> {
     const { sinks } = this.options;
-    await Promise.all(
-      externalCalls.map(async ({ ifaceName, fnName, args }) => {
-        const dep = sinks?.[ifaceName]?.[fnName];
-        if (dep === undefined) {
-          this.log.error('Workflow referenced an unregistered external sink', {
-            ifaceName,
-            fnName,
+
+    const filteredCalls = externalCalls
+      // Map sink call to the corresponding sink function definition
+      .map((call) => ({ call, sink: sinks?.[call.ifaceName]?.[call.fnName] }))
+      // Reject calls to undefined sink definitions
+      .filter(({ call: { ifaceName, fnName }, sink }) => {
+        if (sink !== undefined) return true;
+        this.log.error('Workflow referenced an unregistered external sink', {
+          ifaceName,
+          fnName,
+        });
+        return false;
+      })
+      // If appropriate, reject calls to sink functions not configured with `callDuringReplay = true`
+      .filter(({ sink }) => sink?.callDuringReplay || !isReplaying);
+
+    // Make a wrapper function, to make things easier afterward
+    const mappedCalls = filteredCalls.map(({ call, sink }) => ({
+      call: async () => {
+        try {
+          await sink?.fn(call.workflowInfo, ...call.args);
+        } catch (error) {
+          this.log.error('External sink function threw an error', {
+            ifaceName: call.ifaceName,
+            fnName: call.fnName,
+            error,
+            workflowInfo: call.workflowInfo,
           });
-        } else if (dep.callDuringReplay || !isReplaying) {
-          try {
-            await dep.fn(info, ...args);
-          } catch (error) {
-            this.log.error('External sink function threw an error', {
-              ifaceName,
-              fnName,
-              error,
-              workflowInfo: info,
-            });
-          }
         }
-      })
-    );
+      },
+      sink: sink as InjectedSinkFunction<any>,
+    }));
+
+    const pendingPromises: Promise<void>[] = [];
+    for (const { sink, call } of mappedCalls) {
```

Comment:
You still want different sink functions to run concurrently, I would group the sink calls by `ifaceName, fnName` and run each group concurrently.

Reply:
Grouping on both would lose relative ordering. Doesn't seem desirable. I'd be ok grouping in iface name instead. Can move the concurrency node to this level.